src/celeritas/decay/interactor/MuDecayInteractor.hh

```diff
@@ -0,0 +1,246 @@
+/*!
+ * Construct with shared and state data.
+ */
+CELER_FUNCTION
+MuDecayInteractor::MuDecayInteractor(MuDecayData const& shared,
+                                     ParticleTrackView const& particle,
+                                     Real3 const& inc_direction,
+                                     StackAllocator<Secondary>& allocate)
+    : shared_(shared)
+    , inc_energy_(particle.energy())
+    , inc_direction_(inc_direction)
+    , allocate_(allocate)
+{
+    CELER_EXPECT(particle.particle_id() == shared_.ids.mu_minus
+                 || particle.particle_id() == shared_.ids.mu_plus);
+
+    // Define decay channel
+    is_muon_ = (particle.particle_id() == shared_.ids.mu_minus) ? true : false;
+
+    // Set up muon four vector to boost decay to the lab-frame
+    inc_fourvec_ = {inc_direction_ * particle.momentum().value(),
+                    particle.total_energy().value()};
+
+    // Sampling constants
+    sample_max_
+        = real_type{1}
+          + ipow<2>(shared_.electron_mass.value() / shared_.muon_mass.value());
+
+    // Geant4 physics manual defines E_{max} = m_\mu / 2, while the source code
+    // (since v10.2.0 at least) defines E_{max} = m_\mu / 2 - m_e . The source
+    // code implementation leads to a total CM energy of ~104.6 MeV instead of
+    // the expected 105.7 MeV (muon mass), which is achieved by using the
+    // physics manual definition
+    max_energy_ = real_type{0.5} * shared_.muon_mass.value()
+                  - shared_.electron_mass.value();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the muon decay.
+ *
+ * The decay is sampled at the muon rest frame (based on
+ * \c G4MuonDecayChannel::DecayIt ) and the resulting four vector is boosted to
+ * the lab frame using the muon's incident energy and direction.
+ *
+ * Since only the charged lepton is being returned, these steps are ommitted:
+ * - Calculate the angle between charged lepton and electron neutrino.
+ * - Calculate the energies of the neutrinos.
+ * - Calculate the final directions of the neutrinos using the aforementioned
+ * angle and conservation of momentum.
+ * - Sample a spherically uniform direction and rotate all three secondaries
+ * using `EulerRotation`.
+ * - Boost all particles to the lab frame
+ */
+template<class Engine>
+CELER_FUNCTION Interaction MuDecayInteractor::operator()(Engine& rng)
+{
+    // Allocate secondaries
+    Secondary* secondaries = allocate_(1);
+    if (secondaries == nullptr)
+    {
+        // Failed to allocate secondaries
+        return Interaction::from_failure();
+    }
+
+    real_type electron_energy_frac{};
+    real_type electron_nu_energy_frac{};
+    do
+    {
+        do
+        {
+            electron_nu_energy_frac = UniformRealDist(0, sample_max_)(rng);
+        } while (generate_canonical(rng)
+                 > electron_nu_energy_frac
+                       * (real_type{1} - electron_nu_energy_frac));
```

Comment:
So this is a weird detail that happens in the sampler that doesn't allow me to use any of the standard helper classes (that have reasonable assumptions, such as asserting that the acceptance value is non-negative). Here is the problem:

The `electron_nu_energy_frac` range is `[0, sample_max_)`, which is `[0, 1.000023)`. Now, the `f_` value set in the `RejectionSampler` is `f_ = electron_nu_energy_frac * (1 - electron_nu_energy_frac)`. So that means that for the upper limit (`sample_max_ = 1.000023`), `f_ < 0`, which leads to an assertion error.

Reply:
I agree we should try to refactor this. Do you know the underlying PDFs to be sampled? It would be really good to explicitly document those as well.

Reply:
I also don't think sampling the energy fraction in [0, `sample_max`) makes sense... won't that be equivalent to sampling it in [0, 1), just slightly less efficient?

Reply:
Some thoughts on how this might be refactored: it looks like the inner rejection loop samples the $\nu_e$ energy fraction from the PDF $f(x) = 1/6 x (1 - x)$, $x \in [0, 1)$. This has the CDF $F(x) = 3x^2 - 2x^3$, so to use the inverse transform method we'd have to invert that cubic function.

This PDF is also the beta distribution $\text{B}(2, 2)$. To [sample from a beta distribution](https://en.wikipedia.org/wiki/Beta_distribution#Random_variate_generation) $\text{B}(\alpha, \beta)$, we can draw two independent samples from a gamma distribution, $X \sim \Gamma(\alpha, \theta)$ and $Y \sim \Gamma(\beta, \theta)$: then $\frac{X}{X + Y} \sim \text{B}(\alpha, \beta)$.

Our `GammaDistribution` implements the algorithm described in [Marsaglia and Tsang](https://doi.org/10.1145/358407.358414) which uses a rejection method with a very high (>95%) acceptance rate, but also has nested loops and draw samples from a normal distribution. For this simple case where we'd just need samples from $\Gamma(2, 1)$, we could use [this technique](https://en.wikipedia.org/wiki/Gamma_distribution#Random_variate_generation) which would just require generating two exponentially distributed RVs: $-\sum_{k=1}^n \ln U_k \sim \Gamma(n, 1)$.

I think this would at least be more efficient than the current rejection method.

Reply:
Sorry, just fixed a typo (the PDF normalization is 6,  not 1/6) but yeah g(x) = 3/2 would also be good enough (3 random samples on average).